error: expected initializer before ‘<’ token

### Description

When using the single include header (https://github.com/nlohmann/json/blob/develop/single_include/nlohmann/json.hpp) I get the following error during compilation
`error: expected initializer before ‘<’ token`
in line 5323. 
Sorry if it's an easy fix, I don't have much experience with C++.
I'm using g++ together with the latest CUDA compiler nvcc under WSL.
Edit: does not work using C++20, C++17 (probably below as well) compiles flawlessly

### Reproduction steps

Include "json.hpp" in any header file

### Expected vs. actual results

Expected: Successful compilation
Actual: No compilation

### Minimal code example

_No response_

### Error messages

_No response_

### Compiler and operating system

g++ (Ubuntu 12.1.0-2ubuntu1~22.04) 12.1.0 | Cuda compilation tools, release 12.0, V12.0.76 Build cuda_12.0.r12.0/compiler.31968024_0

### Library version

version 3.11.2

### Validation

- [X] The bug also occurs if the latest version from the [`develop`](https://github.com/nlohmann/json/tree/develop) branch is used.
- [ ] I can successfully [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests).

Can confirm the bug for
- Ubuntu 22.04
- g++12
- cuda 12
- CMake 3.25.2

```
build-debug/_deps/json-src/include/nlohmann/detail/iterators/iteration_proxy.hpp:241:22: error: expected initializer before ‘<’ token
  241 |     inline constexpr bool ::std::ranges::enable_borrowed_range<::nlohmann::detail::iteration_proxy<IteratorType>> = true;
```

No error when the library is compiled with c++17

Yep, the error also occurs on the development branch (commit ac0133ea89a8fd730a9cc9649c5226a9995dc3e1):
```
.../include/nlohmann/detail/iterators/iteration_proxy.hpp:234:22: error: expected initializer before '<' token
  234 |     inline constexpr bool ::std::ranges::enable_borrowed_range<::nlohmann::detail::iteration_proxy<IteratorType>> = true;
      |                      ^
```

Compilers are:
```
-- The CXX compiler identification is GNU 10.5.0                                                                                                            
-- The C compiler identification is GNU 10.5.0  
-- The CUDA compiler identification is NVIDIA 12.0.140                                                                                                      
```
And the error was triggered by an NVCC call

Alright, the reproducer is simple enough:

```cpp
// reproducer.cpp

#include <nlohmann/json.hpp>

int main(){}
```

```bash
/nix/store/7dzdpnzkypv9ywhg9i5jh9ljxrxdv96s-cuda-merged-12.0/bin/nvcc \
    -forward-unknown-to-host-compiler \
    -ccbin=/nix/store/8ssrd7ngjhzhcp7ijnd05v094hkayh6r-gcc-wrapper-10.5.0/bin/g++ \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_CXX_COMPILER_ID=GNU \
    -DCMAKE_CXX_COMPILER_VERSION=10.5.0 \
    -DCMAKE_SYSTEM=Linux-5.15.0-140-generic \
    -DCMAKE_SYSTEM_PROCESSOR=x86_64 \
    -DCMAKE_VERSION=3.30.5 \
    -D_FILE_OFFSET_BITS=64 \
    -D_GNU_SOURCE \
    -D_LARGEFILE_SOURCE \
    -D_POSIX_C_SOURCE=200809L \
    -isystem /nix/store/mbk7c881hdqw0jv4bfa6hjqkrl7rdsy2-cuda_nvcc-12.0.140/include \
    -O3 \
    -DNDEBUG \
    -std=c++20 "--generate-code=arch=compute_52,code=[compute_52,sm_52]" \
    --extended-lambda \
    --expt-relaxed-constexpr \
    --display-error-number \
    -MD \
    -x cu \
    -c ./reproducer.cpp \
    -o ./out.a
```
Outputs:
```
/nix/store/0nbmqd7s6vfjlqq0p9xcpm7nwy4mnhnh-nlohmann_json-ac0133ea89a8fd730a9cc9649c5226a9995dc3e1/include/nlohmann/detail/iterators/iteration_proxy.hpp:234:22: error: expected initializer before '<' token
  234 |     inline constexpr bool ::std::ranges::enable_borrowed_range<::nlohmann::detail::iteration_proxy<IteratorType>> = true;
      |
```

(I used Nix for setting up a local environment that would imitate the CI environment where I observed the error, but from the linked CI runs you can see that this probably makes no big difference.)

--- a/.github/workflows/ubuntu.yml
+++ b/.github/workflows/ubuntu.yml
@@ -237,10 +237,12 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        # matrix to empirically pin down which nvcc/CUDA versions reproduce #3907
-        # (a c++20 parse error in iteration_proxy.hpp's enable_borrowed_range) before
-        # attempting a source-level fix; 11.8 predates CUDA's C++20 support entirely.
-        cuda: ['11.8.0', '12.0.1', '12.1.1', '12.2.2', '12.3.2', '12.4.1', '12.5.1', '12.6.3']
+        # 11.8.0: newest pre-C++20 CUDA release, exercises the C++17 fallback
+        # path (tests/cuda_example/CMakeLists.txt picks the standard per nvcc
+        # version); 12.1.1: permanent regression guard for #3907 (nvcc 12.0/12.1
+        # choke on enable_borrowed_range at C++20, fixed in 12.2); 12.6.3: recent
+        # CUDA/C++20 coverage.
+        cuda: ['11.8.0', '12.1.1', '12.6.3']
     container: nvidia/cuda:${{ matrix.cuda }}-devel-ubuntu22.04
     steps:
       - uses: actions/checkout@9c091bb21b7c1c1d1991bb908d89e4e9dddfe3e0 # v7.0.0

--- a/.github/workflows/ubuntu.yml
+++ b/.github/workflows/ubuntu.yml
@@ -234,11 +234,20 @@ jobs:
 
   ci_cuda_example:
     runs-on: ubuntu-latest
-    container: ghcr.io/nlohmann/json-ci:v2.4.0
+    strategy:
+      fail-fast: false
+      matrix:
+        # matrix to empirically pin down which nvcc/CUDA versions reproduce #3907
+        # (a c++20 parse error in iteration_proxy.hpp's enable_borrowed_range) before
+        # attempting a source-level fix; 11.8 predates CUDA's C++20 support entirely.
+        cuda: ['11.8.0', '12.0.1', '12.1.1', '12.2.2', '12.3.2', '12.4.1', '12.5.1', '12.6.3']
+    container: nvidia/cuda:${{ matrix.cuda }}-devel-ubuntu22.04
     steps:
       - uses: actions/checkout@9c091bb21b7c1c1d1991bb908d89e4e9dddfe3e0 # v7.0.0
         with:
           persist-credentials: false
+      - name: Get latest CMake and ninja
+        uses: lukka/get-cmake@f5b8fbb4d77cec1acc5a5f9f0df4beffaf5d98d9 # v4.3.4
       - name: Run CMake
         run: cmake -S . -B build -DJSON_CI=On
       - name: Build

--- a/cmake/ci.cmake
+++ b/cmake/ci.cmake
@@ -669,7 +669,6 @@ add_custom_target(ci_test_compiler_default
 add_custom_target(ci_cuda_example
     COMMAND ${CMAKE_COMMAND}
         -DCMAKE_BUILD_TYPE=Debug -GNinja
-        -DCMAKE_CUDA_HOST_COMPILER=g++-8
         -S${PROJECT_SOURCE_DIR}/tests/cuda_example -B${PROJECT_BINARY_DIR}/build_cuda_example
     COMMAND ${CMAKE_COMMAND} --build ${PROJECT_BINARY_DIR}/build_cuda_example
 )

--- a/docs/mkdocs/docs/community/quality_assurance.md
+++ b/docs/mkdocs/docs/community/quality_assurance.md
@@ -62,7 +62,9 @@ violations will result in a failed build.
         | Clang 20.1.1                                 | x86_64       | Ubuntu 22.04.1 LTS                | GitHub    |
         | Clang 20.1.8 with GNU-like command-line      | x86_64       | Windows Server 2022 (Build 20348) | GitHub    |
         | Clang 21.1.8                                 | x86_64       | Ubuntu 22.04.1 LTS                | GitHub    |
-        | CUDA 11.0.221 (nvcc)                         | x86_64       | Ubuntu 20.04 LTS                  | GitHub    |
+        | CUDA 11.8.0 (nvcc)                           | x86_64       | Ubuntu 22.04 LTS                  | GitHub    |
+        | CUDA 12.1.1 (nvcc)                           | x86_64       | Ubuntu 22.04 LTS                  | GitHub    |
+        | CUDA 12.6.3 (nvcc)                           | x86_64       | Ubuntu 22.04 LTS                  | GitHub    |
         | Emscripten 4.0.6                             | x86_64       | Ubuntu 22.04.1 LTS                | GitHub    |
         | GNU 4.8.5                                    | x86_64       | Ubuntu 22.04.1 LTS                | GitHub    |
         | GNU 4.9.3                                    | x86_64       | Ubuntu 22.04.1 LTS                | GitHub    |

--- a/include/nlohmann/detail/macro_scope.hpp
+++ b/include/nlohmann/detail/macro_scope.hpp
@@ -146,6 +146,11 @@
         #define JSON_HAS_RANGES 0
     #elif defined(_LIBCPP_VERSION) && _LIBCPP_VERSION < 160000
         #define JSON_HAS_RANGES 0
+        // nvcc CUDA 12.0/12.1 chokes on the enable_borrowed_range variable-template
+        // syntax when compiling as CUDA source; fixed in CUDA 12.2 (issue #3907)
+    #elif defined(__CUDACC__) && defined(__CUDACC_VER_MAJOR__) && __CUDACC_VER_MAJOR__ == 12 \
+        && defined(__CUDACC_VER_MINOR__) && (__CUDACC_VER_MINOR__ == 0 || __CUDACC_VER_MINOR__ == 1)
+        #define JSON_HAS_RANGES 0
     #elif defined(__cpp_lib_ranges)
         #define JSON_HAS_RANGES 1
     #else

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -2520,6 +2520,11 @@ JSON_HEDLEY_DIAGNOSTIC_POP
         #define JSON_HAS_RANGES 0
     #elif defined(_LIBCPP_VERSION) && _LIBCPP_VERSION < 160000
         #define JSON_HAS_RANGES 0
+        // nvcc CUDA 12.0/12.1 chokes on the enable_borrowed_range variable-template
+        // syntax when compiling as CUDA source; fixed in CUDA 12.2 (issue #3907)
+    #elif defined(__CUDACC__) && defined(__CUDACC_VER_MAJOR__) && __CUDACC_VER_MAJOR__ == 12 \
+        && defined(__CUDACC_VER_MINOR__) && (__CUDACC_VER_MINOR__ == 0 || __CUDACC_VER_MINOR__ == 1)
+        #define JSON_HAS_RANGES 0
     #elif defined(__cpp_lib_ranges)
         #define JSON_HAS_RANGES 1
     #else

--- a/tests/cuda_example/CMakeLists.txt
+++ b/tests/cuda_example/CMakeLists.txt
@@ -3,7 +3,18 @@ project(json_cuda LANGUAGES CUDA)
 
 add_executable(json_cuda json_cuda.cu)
 target_include_directories(json_cuda PRIVATE ../../include)
-target_compile_features(json_cuda PUBLIC cuda_std_20)
+
+# nvcc added C++20 support in CUDA 12.0 and C++17 in CUDA 11.0; pick the
+# newest standard the detected compiler actually supports (see #3907)
+# instead of hard-requiring one standard for every CUDA version.
+if(CMAKE_CUDA_COMPILER_VERSION VERSION_GREATER_EQUAL 12.0)
+    set(json_cuda_std 20)
+elseif(CMAKE_CUDA_COMPILER_VERSION VERSION_GREATER_EQUAL 11.0)
+    set(json_cuda_std 17)
+else()
+    set(json_cuda_std 11)
+endif()
+target_compile_features(json_cuda PUBLIC cuda_std_${json_cuda_std})
 set_target_properties(json_cuda PROPERTIES
     CUDA_EXTENSIONS OFF
     CUDA_STANDARD_REQUIRED ON

--- a/tests/cuda_example/CMakeLists.txt
+++ b/tests/cuda_example/CMakeLists.txt
@@ -3,7 +3,7 @@ project(json_cuda LANGUAGES CUDA)
 
 add_executable(json_cuda json_cuda.cu)
 target_include_directories(json_cuda PRIVATE ../../include)
-target_compile_features(json_cuda PUBLIC cuda_std_11)
+target_compile_features(json_cuda PUBLIC cuda_std_20)
 set_target_properties(json_cuda PROPERTIES
     CUDA_EXTENSIONS OFF
     CUDA_STANDARD_REQUIRED ON

--- a/tests/cuda_example/json_cuda.cu
+++ b/tests/cuda_example/json_cuda.cu
@@ -24,8 +24,10 @@ int main()
     nlohmann::json a = {1, 2, 3};
     nlohmann::json b = {1, 2, 3};
     static_cast<void>(a == b);
+#if JSON_HAS_THREE_WAY_COMPARISON
     static_cast<void>(a <=> b); // *NOPAD*
     static_cast<void>(a <=> 1); // *NOPAD*
+#endif
     for (const auto& element : a)
     {
         static_cast<void>(element);

--- a/tests/cuda_example/json_cuda.cu
+++ b/tests/cuda_example/json_cuda.cu
@@ -16,4 +16,18 @@ int main()
     // regression for #3013 (ordered_json::reset() compile error with nvcc)
     nlohmann::ordered_json metadata;
     metadata.erase("key");
+
+    // exercise comparisons (operator==/operator<=>, gated by
+    // JSON_HAS_THREE_WAY_COMPARISON, independent of JSON_HAS_RANGES) and
+    // range-based iteration (exercises iteration_proxy/ranges machinery
+    // beyond just the enable_borrowed_range specialization) — see #3907
+    nlohmann::json a = {1, 2, 3};
+    nlohmann::json b = {1, 2, 3};
+    static_cast<void>(a == b);
+    static_cast<void>(a <=> b); // *NOPAD*
+    static_cast<void>(a <=> 1); // *NOPAD*
+    for (const auto& element : a)
+    {
+        static_cast<void>(element);
+    }
 }

